### PR TITLE
fix: [Bug] Pipeline parameters (ParameterInteger, ParameterString) fail in ModelTrain (5504)

### PR DESCRIPTION
**Pull request by aviruthen**
## Description

No response from agent

## Related Issue

Related issue: 5504

## Changes Made

- `sagemaker-train/src/sagemaker/train/utils.py`
- `sagemaker-train/src/sagemaker/train/model_trainer.py`
- `sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py`

## AI-Generated PR

This PR was automatically generated by the PySDK Issue Agent.
- Confidence score: 85%
- Classification: bug
- SDK version target: V3

## Merge Checklist

- [x] Changes are backward compatible
- [x] Commit message follows `prefix: description` format
- [x] Unit tests added/updated
- [ ] Integration tests added (if applicable)
- [ ] Documentation updated (if applicable)

### CODE REVIEW

File: sagemaker-train/src/sagemaker/train/model_trainer.py [hunk 1]
```diff
@@ -410,14 +410,16 @@ def __del__(self):
                 self._temp_code_dir.cleanup()
 
     def _validate_training_image_and_algorithm_name(
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/model_trainer.py:412]: **Don't remove type annotations — update them instead.** Per SDK coding standards (PEP 484), all public methods MUST have type annotations. Rather than stripping the annotations, update them to accept the union type:

```python
from sagemaker.core.helper.pipeline_variable import PipelineVariable

def _validate_training_image_and_algorithm_name(
    self, training_image: str | PipelineVariable | None, algorithm_name: str | PipelineVariable | None
):
```

This preserves type safety and documents the intended usage.

File: sagemaker-train/src/sagemaker/train/utils.py [hunk 2]
```diff
@@ -142,7 +142,7 @@ def _get_unique_name(base, max_length=63):
     return unique_name
 
 
-def _get_repo_name_from_image(image: str) -> str:
+def _get_repo_name_from_image(image) -> str:
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/utils.py]: **Same issue — don't remove the type annotation, update it:**

```python
def _get_repo_name_from_image(image: str | PipelineVariable) -> str | None:
```

Note the return type should also be updated to `str | None` since you now return `None` for PipelineVariable inputs.

File: sagemaker-train/src/sagemaker/train/utils.py [hunk 3]
```diff
@@ -152,11 +152,15 @@ def _get_repo_name_from_image(image: str) -> str:
...
     Args:
-        image (str): The image URI
+        image: The image URI (str or PipelineVariable)
 
     Returns:
-        str: The repository name
+        str: The repository name, or None if image is a PipelineVariable
     """
+    from sagemaker.core.helper.pipeline_variable import PipelineVariable
+
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/utils.py]: **Move the import to the top of the module** (or at minimum to the top of the function). Inline imports inside functions are acceptable for avoiding circular dependencies, but please add a comment explaining why it's done here:

```python
# Import here to avoid circular dependency
from sagemaker.core.helper.pipeline_variable import PipelineVariable
```

Also, is there actually a circular dependency risk? If not, this import should be at the module level with other imports.

File: sagemaker-train/src/sagemaker/train/utils.py [hunk 4]
```diff
@@ -152,11 +152,15 @@ def _get_repo_name_from_image(image: str) -> str:
...
+    if isinstance(image, PipelineVariable):
+        return None
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/utils.py:161]: **Returning `None` here could cause downstream `AttributeError` or `TypeError` exceptions** wherever the return value of `_get_repo_name_from_image` is used (e.g., string concatenation, further splitting). Have you traced all callers to ensure they handle `None` gracefully? If a caller does `_get_repo_name_from_image(image) + '-suffix'`, this will crash. Please verify and add `None` checks at call sites, or consider returning a sentinel string like `'pipeline-variable'` instead.

File: sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py [hunk 5]
```diff
@@ -33,7 +33,7 @@
     StoppingCondition,
     OutputDataConfig,
 )
-from sagemaker.train.defaults import DEFAULT_INSTANCE_TYPE
+from sagemaker.train.defaults import DEFAULT_INSTANCE_TYPE, TrainDefaults
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py]: **`TrainDefaults` is imported but never used in the diff.** Is this import used elsewhere in the file (not shown in the diff)? If not, this is an unused import that will fail linting. If it IS used in existing code not shown in the diff, please disregard this comment.

Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py]: **The test changes seem insufficient for the scope of the fix.** The PR modifies validation logic in `_validate_training_image_and_algorithm_name` and adds `PipelineVariable` handling in `_get_repo_name_from_image`, but the test diff only shows an import change. Where are the new test cases that:
1. Pass a `PipelineVariable` as `training_image` and verify validation passes?
2. Pass a `PipelineVariable` to `_get_repo_name_from_image` and verify it returns `None`?
3. Verify that both `PipelineVariable` for `training_image` AND `algorithm_name` still raises `ValueError`?
4. Verify that `None` for both still raises `ValueError`?

Please add explicit unit tests for the changed behavior. Target >90% coverage per SDK standards.

File: sagemaker-train/src/sagemaker/train/utils.py [hunk 6]
```diff
@@ -142,7 +142,7 @@ def _get_unique_name(base, max_length=63):
     return unique_name
 
 
-def _get_repo_name_from_image(image: str) -> str:
+def _get_repo_name_from_image(image: "str | PipelineVariable") -> "str | None":
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/utils.py:145]: The type annotation uses PEP 604 string-quoted syntax (`"str | PipelineVariable"`) for the parameter but the module doesn't use `from __future__ import annotations`. While this works at runtime due to quoting, it's inconsistent with the return type which is also quoted. Per SDK conventions, prefer adding `from __future__ import annotations` at the top of the module and using unquoted `str | PipelineVariable` and `str | None`, or use `Union[str, PipelineVariable]` / `Optional[str]` consistently with the rest of the codebase (as done in `model_trainer.py`).

File: sagemaker-train/src/sagemaker/train/utils.py [hunk 7]
```diff
@@ -24,9 +24,9 @@
 from typing import Literal, Any
 
 from sagemaker.core.helper.session_helper import Session
+from sagemaker.core.helper.pipeline_variable import PipelineVariable
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/utils.py:27]: The import was changed from `sagemaker.core.workflow.parameters.PipelineVariable` to `sagemaker.core.helper.pipeline_variable.PipelineVariable`. Please verify these are the same class (or that the helper version is the correct base class that `ParameterString`/`ParameterInteger` inherit from). If `sagemaker.core.workflow.parameters.PipelineVariable` is a re-export of the same class, this is fine, but if they're different classes, the `isinstance` check on line 163 might not catch all pipeline variable types.

File: sagemaker-train/src/sagemaker/train/model_trainer.py [hunk 8]
```diff
@@ -410,14 +410,18 @@ def __del__(self):
...
-        self, training_image: Optional[str], algorithm_name: Optional[str]
+        self,
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/model_trainer.py:413]: Good fix switching from truthiness (`not training_image`) to identity (`is not None`) checks. The truthiness check would evaluate a PipelineVariable object as truthy but could also fail for empty strings. The `is not None` approach is more correct and explicit.

Minor: The type annotation uses `Union[str, PipelineVariable, None]` — consider whether `Optional[Union[str, PipelineVariable]]` or `str | PipelineVariable | None` (with `from __future__ import annotations`) would be more consistent with the SDK's preference for newer syntax in new code.

File: sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py [hunk 9]
```diff
@@ -176,3 +177,116 @@ def test_training_image_rejects_invalid_type(self):
...
+
+    def test_safe_serialize_dict_returns_json(self):
+        """safe_serialize should JSON-encode dicts."""
+        result = safe_serialize({"key": "value"})
+        assert result == '{"key": "value"}'
+
+    def test_safe_serialize_parameter_string_returns_variable(self):
+        """safe_serialize should return ParameterString as-is."""
+        param = ParameterString(name="MyParam", default_value="val")
+        result = safe_serialize(param)
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py:291]: The `TestSafeSerialize` tests are good additions, but I don't see any corresponding changes to the `safe_serialize` function in `utils.py` in this diff. If `safe_serialize` already handles `PipelineVariable` correctly, these tests are documenting existing behavior (which is fine). But if the bug also affects `safe_serialize` and it needs a fix, that change appears to be missing from this PR. Could you confirm that `safe_serialize` already has the PipelineVariable guard?

File: sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py [hunk 10]
```diff
@@ -176,3 +177,116 @@ def test_training_image_rejects_invalid_type(self):
                 stopping_condition=DEFAULT_STOPPING,
                 output_data_config=DEFAULT_OUTPUT,
             )
+
+
+class TestValidateTrainingImageAndAlgorithmName:
+    """Tests for _validate_training_image_and_algorithm_name with PipelineVariable."""
+
+    def test_pipeline_variable_training_image_passes_validation(self):
+        """PipelineVariable as training_image should pass validation."""
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py:186]: Good test coverage for the validation fix. Consider also adding a test case where `training_image` is a `ParameterString` with an empty string default value (`default_value=""`) to verify the `is not None` check doesn't regress for edge cases where the resolved value might be empty.
